text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,629 @@
+```
+
+This might generate a graph like the following (the following sections
+will describe how this graph is constructed). Note that this is not a
+complete graph, it only shows the data needed to produce `MIR(foo)`.
+
+```
+BODY(foo) ----------------------------> TYPECK(foo) --> MIR(foo)
+                                          ^ ^ ^ ^         |
+SIG(foo) ----> COLLECT(foo)               | | | |         |
+                 |                        | | | |         v
+                 +--> ITEM_TYPE(foo) -----+ | | |      LLVM(foo)
+                 +--> PREDICATES(foo) ------+ | |         |
+                                              | |         |
+SIG(bar) ----> COLLECT(bar)                   | |         v
+                 |                            | |     OBJECT(foo)
+                 +--> ITEM_TYPE(bar) ---------+ |
+                 +--> PREDICATES(bar) ----------+
+```
+
+As you can see, this graph indicates that if the signature of either
+function changes, we will need to rebuild the MIR for `foo`. But there
+is no path from the body of `bar` to the MIR for foo, so changes there
+need not trigger a rebuild (we are assuming here that `bar` is not
+inlined into `foo`; see the [section on optimizations](#optimization)
+for more details on how to handle those sorts of dependencies).
+
+### Building the graph
+
+It is very important the dependency graph contain *all* edges. If any
+edges are missing, it will mean that we will get inconsistent builds,
+where something should have been rebuilt what was not. Hand-coding a
+graph like this, therefore, is probably not the best choice -- we
+might get it right at first, but it's easy to for such a setup to fall
+out of sync as the code is edited. (For example, if a new table is
+added, or a function starts reading data that it didn't before.)
+
+Another consideration is compiler plugins. At present, of course, we
+don't have a stable API for such plugins, but eventually we'd like to
+support a rich family of them, and they may want to participate in the
+incremental compilation system as well. So we need to have an idea of
+what data a plugin accesses and modifies, and for what purpose.
+
+The basic strategy then is to build the graph dynamically with an API
+that looks something like this:
+
+- `push_procedure(procedure_node)`
+- `pop_procedure(procedure_node)`
+- `read_from(data_node)`
+- `write_to(data_node)`
+
+Here, the `procedure_node` arguments are one of the procedure labels
+above (like `COLLECT(X)`), and the `data_node` arguments are either
+HIR or IR nodes (e.g., `SIG(X)`, `MIR(X)`).
+
+The idea is that we maintain for each thread a stack of active
+procedures. When `push_procedure` is called, a new entry is pushed
+onto that stack, and when `pop_procedure` is called, an entry is
+popped. When `read_from(D)` is called, we add an edge from `D` to the
+top of the stack (it is an error if the stack is empty). Similarly,
+`write_to(D)` adds an edge from the top of the stack to `D`.
```

Comment:
Why are procedures on a stack? Is this conceptually sound? Are procedure executions really always cleanly nested? Where do multiple active procedures even occur? (monomorphization maybe?)
In general, I think, as you write later, it would be best to try and enforce only having one active procedure at a time.